src/pybind/mgr/ansible/module.py

```diff
@@ -673,6 +694,120 @@ def remove_host(self, host):
 
         return ARSChangeOperation(self.ar_client, self.log, operations)
 
+# JMO --------------------------------------------------------------------------
+    def add_stateless_service(self, service_type, spec):
+        """ Add a stateless service in the cluster
+
+        : service_type: Kind of service (nfs, rgw, mds)
+        : spec        : an Orchestrator.StatelessServiceSpec object
+
+        : returns     : Completion object
+        """
+
+        # Check service_type is supported
+        if service_type not in ["rgw"]:
+            raise orchestrator.OrchestratorError(
+                "{} service not supported".format(service_type))
+
+        # Add the hosts to the inventory in the right group
+        group = "{}s".format(service_type)
+        hosts = spec.service_spec.hosts
+        if not hosts:
+            raise orchestrator.OrchestratorError("No hosts provided."\
```

Comment:
can we move this check to `StatelessServiceSpec`?

Reply:
+1 I think validating internal properties is a concern of the class itself (or a delegated class), but not for the calling method.

If other orchestrators don't require hosts to be defined, or Ansible brings in specific validation/adaptation logic, in that case I'd go for an [adapter pattern](https://www.giacomodebidda.com/adapter-pattern-in-python/).

Reply:
The check of hosts is specific to Ansible Orchestrator. So probably not a good idea ( Rook probably does not need that).

Maybe i do not understand you @epuertat , but in this case the use of the adapter pattern will force to select one kind of adapter in the clients... so clients will need to know what is the backend orchestrator .. and that is not ok.

Reply:
This is `ansible/module.py` so I assume everything is Ansible-only here. The adapter class happens here.

This is just a matter of taste, but when mixing methods (as required by CLI commands) with objects, I usually prefer to have very lightweight methods, with the minimal logic to handle I/O, while moving any validation to a class (in this case, as Ansible requires `hosts` to be defined, this would fall under the umbrella of the `AnsibleSpecAdapter`).

Why? If you need to create more CLI commands for other kind of services for Ansible, and, for example, all of them should fail when no hosts are defined, this looks like something you could easily move to a common class, rather than in every new CLI command you create.